services/port_manager/src/main/java/com/futurewei/alcor/portmanager/restwrap/IpAddressRestWrap.java

```diff
@@ -0,0 +1,154 @@
+import com.futurewei.alcor.portmanager.rollback.ReleaseIpAddrRollback;
+import com.futurewei.alcor.web.rest.SubnetRest;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+
+public class IpAddressRestWrap {
+    private IpAddressRest ipAddressRest;
+    private SubnetRest subnetRest;
+    private Stack<PortStateRollback> rollbacks;
+    private String projectId;
+
+    public IpAddressRestWrap(Stack<PortStateRollback> rollbacks, String projectId) {
+        ipAddressRest = BeanUtil.getBean(IpAddressRest.class);
+        subnetRest = BeanUtil.getBean(SubnetRest.class);
+        this.rollbacks = rollbacks;
+        this.projectId = projectId;
+    }
+
+    private int getIpVersion(String ipAddress) throws Exception {
+        if (Ipv4AddrUtil.formatCheck(ipAddress)) {
+            return IpVersion.IPV4.getVersion();
+        } else if (Ipv6AddrUtil.formatCheck(ipAddress)) {
+            return IpVersion.IPV6.getVersion();
+        } else {
+            throw new IpAddrInvalidException();
+        }
+    }
+
+    private String getRangeIdBySubnetId(String subnetId, int ipVersion) throws Exception {
+        SubnetStateJson subnetStateJson = subnetRest.getSubnetState(projectId, subnetId);
+        if (subnetStateJson == null || subnetStateJson.getSubnet() == null) {
+            throw new VerifySubnetException();
+        }
+
+        if (IpVersion.IPV4.getVersion() == ipVersion) {
+            return subnetStateJson.getSubnet().getIpV4RangeId();
+        } else if (IpVersion.IPV6.getVersion() == ipVersion) {
+            return subnetStateJson.getSubnet().getIpV4RangeId();
+        }
+
+        throw new IpVersionInvalidException();
+    }
+
+    private void addIpAddrRollback(AbstractIpAddrRollback rollback, IpAddrRequest ipAddr) {
+        if (rollback instanceof AllocateIpAddrRollback) {
+            rollback.putAllocatedIpAddress(ipAddr);
+        } else {
+            rollback.putReleasedIpAddress(ipAddr);
+        }
+
+        rollbacks.push(rollback);
+    }
+
+    public List<IpAddrRequest> allocateIpAddress(Object args) throws Exception {
+        List<IpAddrRequest> ipAddrRequests = new ArrayList<>();
+        PortState portState = (PortState)args;
+
+        IpAddrRequest result = ipAddressRest.allocateIpAddress(IpVersion.IPV4,
+                portState.getVpcId(), null, null);
+
+        List<PortState.FixedIp> fixedIps = new ArrayList<>();
+        PortState.FixedIp fixedIp = new PortState.FixedIp(result.getSubnetId(), result.getIp());
+
+        fixedIps.add(fixedIp);
+        portState.setFixedIps(fixedIps);
+
+        addIpAddrRollback(new AllocateIpAddrRollback(ipAddressRest), result);
+
+        ipAddrRequests.add(result);
+
+        return ipAddrRequests;
+    }
+
+    public List<IpAddrRequest> verifyIpAddresses(Object args) throws Exception {
+        List<IpAddrRequest> ipAddrRequests = new ArrayList<>();
+        List<PortState.FixedIp> fixedIps = (List<PortState.FixedIp>)args;
```

Comment:
In general, direct type casting is not recommended for downcasting.

Could you try cast() and isInstance() methods for safer downcasting? 

A reference: https://www.baeldung.com/java-type-casting

Reply:
Don't worry, an exception will be thrown if there is a type conversion error, which is exactly what I expected when the type conversion failed.